messages_controller.go

```diff
@@ -64,87 +72,106 @@ func NewMessagesViewController(
 	}
 }
 
-func (c *MessagesViewController) Start() {
+func (c *MessagesViewController) Start() error {
+
 	if c.cancel == nil {
 		c.cancel = make(chan struct{})
+		chats, err := c.messenger.Chats()
+		if err != nil {
+			return err
+		}
+
+		for _, chat := range chats {
+			// Pull latest 10 messages
+			latestMessages, _, err := c.messenger.MessageByChatID(chat.ID, "", 10)
+			if err != nil {
+				return err
+			}
+
+			c.mutex.Lock()
+			for _, m := range latestMessages {
+				c.store[chat.ID] = append(c.store[chat.ID], m)
+			}
+			c.mutex.Unlock()
+		}
+
 		go c.readMessagesLoop()
 	}
+	return nil
 }
 
 func (c *MessagesViewController) readMessagesLoop() {
 	c.done = make(chan struct{})
 	defer close(c.done)
 
-	// TODO: It should be a round buffer instead.
-	// It is a map with chatID as a key and a list of messages.
-	store := make(map[string][]*v1protocol.Message)
-
 	t := time.NewTicker(time.Second)
 	defer t.Stop()
 
 	for {
 		select {
 		case <-t.C:
-			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
-			allLatest, err := c.messenger.RetrieveAll(ctx, protocol.RetrieveLatest)
-			cancel()
+			response, err := c.messenger.RetrieveAll()
 			if err != nil {
 				c.logger.Error("failed to retrieve messages", zap.Error(err))
 				continue
 			}
 
-			c.logger.Debug("received latest messages", zap.Int("count", len(allLatest)))
+			c.logger.Info("received latest messages", zap.Int("count", len(response.Messages)))
 
-			for _, m := range allLatest {
-				store[m.ChatID] = append(store[m.ChatID], m)
+			c.mutex.Lock()
+			for _, m := range response.Messages {
+				c.store[m.LocalChatID] = append(c.store[m.LocalChatID], m)
 			}
 
 			c.onMessages()
 
 			if c.activeChat == nil {
+				c.mutex.Unlock()
 				break
 			}
 
-			var latestForActive []*v1protocol.Message
-			for _, m := range allLatest {
-				if m.ChatID == c.activeChat.ID {
+			var latestForActive []*protocol.Message
+			for _, m := range response.Messages {
+				if m.LocalChatID == c.activeChat.ID {
 					latestForActive = append(latestForActive, m)
 				}
 			}
 
 			if len(latestForActive) == 0 {
+				c.mutex.Unlock()
```

Comment:
Maybe it would be cleaner if we moved this case into a separate function, and then we could leverage `defer` for the mutex unlocking?

Reply:
sounds good